crypto/src/main/java/org/hyperledger/besu/crypto/SECP256K1.java

```diff
@@ -308,6 +336,99 @@ public static Bytes calculateECIESKeyAgreement(final PrivateKey privKey, final P
+        LibSecp256k1.CONTEXT, compactSig, recId, signature);
+    compactSig.flip();
+    final byte[] sig = compactSig.array();
+
+    // wrap in signature object
+    final Bytes32 r = Bytes32.wrap(sig, 0);
+    final Bytes32 s = Bytes32.wrap(sig, 32);
+    return Signature.create(
+        r.toUnsignedBigInteger(), s.toUnsignedBigInteger(), (byte) recId.getValue());
+  }
+
+  private static boolean verifyNative(
+      final Bytes data, final Signature signature, final PublicKey pub) {
+
+    // translate signature
+    final LibSecp256k1.secp256k1_ecdsa_signature _signature = new secp256k1_ecdsa_signature();
+    if (LibSecp256k1.INSTANCE.secp256k1_ecdsa_signature_parse_compact(
+            LibSecp256k1.CONTEXT, _signature, signature.encodedBytes().toArrayUnsafe())
+        == 0) {
+      throw new IllegalArgumentException("Could not parse signature");
+    }
+
+    // translate key
+    final LibSecp256k1.secp256k1_pubkey _pub = new secp256k1_pubkey();
+    final Bytes encodedPubKey = Bytes.concatenate(Bytes.of(0x04), pub.getEncodedBytes());
+    if (LibSecp256k1.INSTANCE.secp256k1_ec_pubkey_parse(
+            LibSecp256k1.CONTEXT, _pub, encodedPubKey.toArrayUnsafe(), encodedPubKey.size())
+        == 0) {
+      throw new IllegalArgumentException("Could not parse public key");
+    }
+
+    return LibSecp256k1.INSTANCE.secp256k1_ecdsa_verify(
+            LibSecp256k1.CONTEXT, _signature, data.toArrayUnsafe(), _pub)
+        != 0;
+  }
+
+  private static Optional<PublicKey> recoverFromSignatureNative(
+      final Bytes32 dataHash, final Signature signature) {
+
+    // parse the sig
+    final LibSecp256k1.secp256k1_ecdsa_recoverable_signature parsedSignature =
+        new LibSecp256k1.secp256k1_ecdsa_recoverable_signature();
+    final Bytes encodedSig = signature.encodedBytes();
+    if (LibSecp256k1.INSTANCE.secp256k1_ecdsa_recoverable_signature_parse_compact(
+            LibSecp256k1.CONTEXT,
+            parsedSignature,
+            encodedSig.slice(0, 64).toArrayUnsafe(),
+            encodedSig.get(64))
+        == 0) {
+      throw new IllegalArgumentException("Could not parse signature");
+    }
+
+    // recover the key
+    final LibSecp256k1.secp256k1_pubkey newPubKey = new LibSecp256k1.secp256k1_pubkey();
+    if (LibSecp256k1.INSTANCE.secp256k1_ecdsa_recover(
+            LibSecp256k1.CONTEXT, newPubKey, parsedSignature, dataHash.toArrayUnsafe())
+        == 0) {
+      throw new IllegalArgumentException("Could not recover public key");
+    }
+
+    // parse the key
+    final ByteBuffer recoveredKey = ByteBuffer.allocate(65);
+    final LongByReference keySize = new LongByReference(recoveredKey.limit());
+    LibSecp256k1.INSTANCE.secp256k1_ec_pubkey_serialize(
+        LibSecp256k1.CONTEXT,
+        recoveredKey,
+        keySize,
+        newPubKey,
+        LibSecp256k1.SECP256K1_EC_UNCOMPRESSED);
+
+    return Optional.of(PublicKey.create(Bytes.wrapByteBuffer(recoveredKey).slice(1)));
```

Comment:
Does the native `recoveredKey` return null in the same cases as the java implementation?

Reply:
Legacy APIs.  The old java method had a way to return emptys, but when I try and return empty from any of the methods that return error I get reference tests that break.  I'm not sure why the java method gets away with it.